Core/Core/Inbox/ComposeMessage/ViewModel/ComposeMessageViewModel.swift

```diff
@@ -0,0 +1,192 @@
+        let addressbook = AddressBookAssembly.makeAddressbookViewController(courseID: courseID, recipientDidSelect: selectedRecipient)
+        router.show(addressbook, from: viewController)
+    }
+
+    public func attachmentbuttonDidTap(viewController: WeakViewController) {
+
+    }
+
+    public func removeRecipientButtonDidTap(recipient: SearchRecipient) {
+        recipients.removeAll { $0 == recipient}
+    }
+
+    private func setupOutputBindings() {
+        interactor.state
+            .assign(to: &$state)
+        interactor.courses
+            .assign(to: &$courses)
+        selectedRecipient
+            .compactMap { $0 }
+            .filter { !self.recipients.map { $0.id }.contains($0.id) }
+            .sink { [weak self] in
+                self?.recipients.append($0)
+            }
+            .store(in: &subscriptions)
+    }
+
+    private func messageParameters() -> MessageParameters? {
+        guard let courseID = selectedCourse?.courseId else { return nil }
+        let recipientIDs = recipients.map { $0.id }
+
+        return MessageParameters(
+            subject: subject,
+            body: bodyText,
+            recipientIDs: recipientIDs,
+            context: .course(courseID)
+        )
+    }
+
+    private func showResultDialog(title: String, message: String, completion: (() -> Void)? = nil) {
+        let title = NSLocalizedString(title, comment: "")
+        let message = NSLocalizedString(message, comment: "")
+        let actionTitle = NSLocalizedString("OK", comment: "")
+        let alert = UIAlertController(title: title, message: message, preferredStyle: .alert)
+        let action = UIAlertAction(title: actionTitle, style: .default) { _ in
+            completion?()
+        }
+        alert.addAction(action)
+
+        if let top = AppEnvironment.shared.topViewController {
+            AppEnvironment.shared.router.show(alert, from: top, options: .modal())
+        }
+    }
+
+    private func setupInputBindings(router: Router) {
+        cancelButtonDidTap
+            .sink { [router] viewController in
+                router.dismiss(viewController)
+            }
+            .store(in: &subscriptions)
+        sendButtonDidTap
+            .compactMap { [weak self] viewController -> (WeakViewController, MessageParameters)? in
+                guard let self = self, let params = self.messageParameters() else { return nil }
+                return (viewController, params)
+            }
+            .handleEvents(receiveOutput: { [weak self] (viewController, _) in
+                self?.isSendingMessage = true
+                self?.router.dismiss(viewController)
+            })
+            .flatMap { [interactor] (viewController, params) in
+                interactor
+                    .send(parameters: params)
+                    .map {
+                        viewController
+                    }
+            }
+            .receive(on: RunLoop.main)
```

Comment:
Instead of hard coding `RunLoop.main` we could inject a `scheduler: AnySchedulerOf<DispatchQueue> = .main` to the view model's initializer and then we'd be able to make the unit tests synchronous by injecting `.immediate`.